Data masking docs (#12162)
apollographql/apollo-client: docs/source/data/fragments.mdx

</Note>

See the [API reference](../api/react/hooks#usefragment) for more details on the supported options.

<MinVersion version="3.12.0">

> ## Data masking

Author: Reminder to self: Add a mention of performance characteristics

</MinVersion>

By default, Apollo Client returns all data for all fields defined in a GraphQL operation. As your app grows and scales over time, this can create a high degree of coupling between the components that query your GraphQL data and the component subtree underneath it. [Colocated fragments](#colocating-fragments) reduce the degree of coupling by moving component's data requirements into fragments, however on its own does not entirely eliminate it.

> Let's take a look at an example.

Author: Good call!

  if (allPosts.length === 0) {
    return <div>No posts to display</div>;
  }

  return (
    <div>
      {allPosts.map((post) => (
        <PostDetails key={post.id} post={post} />
      ))}
    </div>
  );
}
```

```jsx title="PostDetails.jsx"
export const POST_DETAILS_FRAGMENT = gql`
  fragment PostDetailsFragment on Post {
    title
    shortDescription
    publishedAt
  }
`;

export default function PostDetails({ post }) {
  return (
    <section>
      <h1>{post.title}</h1>
      <p>{post.shortDescription}</p>
      <p>
        {post.publishedAt ?
          `Published: ${formatDate(post.publishedAt)}`
        : 'Private'}
      </p>
    </section>
  );
}
````

The `Posts` component is responsible for fetching and rendering a list of posts. We loop over each post and render a `PostDetails` component to display details about the post. `PostDetails` uses a colocated fragment to define it's own data requirements necessary to render post details, which is included in the `GetPosts` query. When the `includeUnpublishedPosts` prop is false, the `Posts` component filters out unpublished posts from our list of all posts by checking the `publishedAt` property on the post object.

> This may work great for a while, but consider what happens when we start modifying our `PostDetails` component.

Author: Good to know! Thanks 🙂

We've decided that we no longer want to show the publish date when viewing the list of posts and prefer to display it when viewing an individual post. Let's modify `PostDetails` accordingly.

```jsx title="PostDetails.jsx"
export const POST_DETAILS_FRAGMENT = gql`
  fragment PostDetailsFragment on Post {
    title
    shortDescription
  }
`;

export default function PostDetails({ post }) {
  return (
    <section>
      <h1>{post.title}</h1>
      <p>{post.shortDescription}</p>
    </section>
  );
}
````

We've removed the check for `publishedAt` since we no longer show the publish date. We've also removed the `publishedAt` field from the `PostDetailsFragment` fragment since we no longer use this field in the `PostDetails` component.

Uh oh, we just broke our app! Our `Posts` component no longer shows any posts! Our `Posts` component still depends on `publishedAt`, but because the field was declared in the `PostDetailsFragment` fragment, changes to `PostDetails` resulted in a subtle breakage of our `Posts` component.

> This is an example of an **implicit dependency** between our components. As the application grows in complexity, these types of implicit dependencies can be more and more difficult to track. Imagine if `PostDetails` was a component nested much deeper in our component tree or if this component was used in multiple queries!

Reviewer: ```suggestion
This coupling is an example of an **implicit dependency** between components. As the application grows in complexity, these implicit dependencies can become more and more difficult to track. Imagine if `PostDetails` was a component nested much deeper in our component tree or if multiple queries used it.
```

> Data masking helps us eliminate these types of implicit dependencies by returning only the data declared by the component's query or fragment. This creates more loosely coupled components that are more resistant to change.

Reviewer: Bringing emphasis to the key term.

```suggestion
**Data masking** helps eliminate these types of implicit dependencies by returning only the data declared by the component's query or fragment. As a result, data masking creates more loosely coupled components that are more resistant to change.
```

### Enabling data masking

To enable data masking in Apollo Client, set the `dataMasking` flag in the `ApolloClient` constructor to `true`.

```jsx {2}
const client = new ApolloClient({
  dataMasking: true,
  // ...
});
```

> When `dataMasking` is enabled, fields defined in fragments are hidden from our component. This prevents us from accessing data the component didn't ask for.

Reviewer: What exactly is "our" component? Is it just any components that use fragments?

```suggestion
When `dataMasking` is enabled, fields defined in fragments are hidden from components. This prevents us from accessing data the component didn't ask for.
```

Author: ```suggestion
When `dataMasking` is enabled, fields defined in fragments are hidden from components. This prevents the component from accessing data it didn't ask for.
```

I don't like the `us` in there, so removing that as well 🙂

> Enabling data masking enables masking on all operation types and extends to all request-based APIs, such as `useQuery`, `client.query`, `client.mutate`, etc. Cache APIs, such as `cache.readQuery` and `cache.readFragment` are never masked.

Reviewer: ```suggestion
Enabling data masking applies it to all operation types and all request-based APIs, such as `useQuery`, `client.query`, `client.mutate`, etc. Cache APIs, such as `cache.readQuery` and `cache.readFragment` are never masked.
```

  const { data, loading } = useQuery(GET_POSTS);

  // ...
}
```

Our `GetPosts` query asks for the `posts` field along with an `id` for each post. All other fields are defined in `PostDetailsFragment`. If we were to inspect `data`, we'd see that the only accessible fields are those defined in our query but not the fragment.

```json
{
  "posts": [
    {
      "__typename": "Post",
      "id": "1"
    },
    {
      "__typename": "Post",
      "id": "2"
    }
  ]
}
```

We can access more data by adding fields to our query. Let's fix the previous section's example by adding the `publishedAt` field to our `GetPosts` query so that our `Posts` component can use it.

```jsx {5} title="Posts.jsx"
const GET_POSTS = gql`
  query GetPosts {
    posts {
      id
      publishedAt
      ...PostDetailsFragment
    }
  }

  ${POST_DETAILS_FRAGMENT}
`;
```

Now if we inspect `data`, we'll see that `publishedAt` is available to the `Posts` component.

```json
{
  "posts": [
    {
      "__typename": "Post",
      "publishedAt": "2024-01-01",
      "id": "1"
    },
    {
      "__typename": "Post",
      "publishedAt": null,
      "id": "2"
    }
  ]
}
```

### Reading fragment data

> Now that our `GetPosts` query is masked, we've introduced a problem for our `PostDetails` component. The `post` prop no longer contains the fields from our `PostDetailsFragment` fragment, preventing us from rendering that data!

Reviewer: Trying to limit `!` to reserve for emphasis--feel free to make the call on where it makes most sense to include them! (I try to limit to 2 or 3 max on a page.)

```suggestion
Now that our `GetPosts` query is masked, we've introduced a problem for our `PostDetails` component. The `post` prop no longer contains the fields from our `PostDetailsFragment` fragment, preventing us from rendering that data.
```

    fragment: POST_DETAILS_FRAGMENT,
    from: post,
  });

  // ...
}
```

Now we use the `data` property returned from `useFragment` to render the details from the `post`.

```jsx {17-18} title="PostDetails.jsx"
function PostDetails({ post }) {
  const { data, complete } = useFragment({
    fragment: POST_DETAILS_FRAGMENT,
    from: post,
  });

  // It's a good idea to check the `complete` flag to ensure all data was
  // successfully queried from the cache. This can indicate a potential
  // issue with our cache configuration or parent object when `complete`
  // is `false`.
  if (!complete) {
    return null;
  }

  return (
    <section>
      <h1>{data.title}</h1>
      <p>{data.shortDescription}</p>
    </section>
  )
}
```

<Note>
It's important that the parent query or fragment selects any [`keyFields`](../caching/cache-configuration#customizing-cache-ids) for objects passed to the `from` option. Without this, we'd have no way to identify the object with the cache and the `data` returned from `useFragment` would be incomplete. If you forget to include key fields in the parent object, you will see a warning in the console.
</Note>

> ### Nesting fragments in other fragments

Reviewer: It seems like this could be a topic outside of Data Masking. I'm guessing it's a pattern we only really want to promote with data masking enabled? If so, maybe that's something we want to callout explicitly?

Reviewer: Hmm, I think it's fine wheres it currently is! Might be worth just adding something like this:

```suggestion
### Nesting fragments in other fragments

<Note>

You can nest fragments with or [without data masking](#relevant-anchor), the following section describes how you can use masking in components with `useFragment`.)

</Note>

```

Author: I dig it. Thanks!

</Note>

If we inspect the `data` property returned by `useFragment` in `PostDetails`, we'll see that only the fields included by the `PostDetailsFragment` fragment are a part of the object.

```json
{
  "__typename": "Post",
  "title": "The Amazing Adventures of Data Masking",
  "shortDescription": "In this article we dive into...",
  "topComment": {
    "__typename": "Comment",
    "id": "1"
  }
}
```

Throughout this example, You'll notice that we never touched the `GetPosts` query as a result of this change. Because we included `CommentFragment` with `PostDetailsFragment`, it was added to our query automatically. Colocating fragments like this is a powerful pattern that, when combined with data masking, provide very self-isolated components.

<Tip>
We recommend that parent components only add fragments defined by their directly-rendered children to prevent coupling with more deeply nested components. In this example, the `GetPosts` query did not include the `CommentFragment` fragment directly but rather it relied on the `PostDetails` component to include it with the `PostDetailsFragment` fragment.
</Tip>

### Working with other operation types

> Data masking is not limited to queries but also extends to other operation types. As a rule of thumb, any value that is used to _read_ data from a request-based API is masked. APIs that perform cache updates are never masked.

Reviewer: ```suggestion
Data masking is not limited to queries but also extends to other operation types. As a rule of thumb, any value that is used to _read_ data from a request-based API is masked. APIs that perform cache updates are never masked.

Refer to the code samples below to see what data is masked in mutations and subscriptions.
```

  updateQueries: {
    ExampleQuery: (previous, { mutationResult }) => {
      // mutationResult is unmasked
    }
  }
});

async function runMutation() {
  const { data } = await mutate()

  // data is masked
}
```

```jsx
// data is masked
const { data } = await client.mutate({
  update: (cache, { data }) => {
    // data is unmasked
  },
  refetchQueries: ({ data }) => {
    // data is unmasked
  },
  updateQueries: {
    ExampleQuery: (previous, { mutationResult }) => {
      // mutationResult is unmasked
    }
  }
});
```

#### Subscriptions

```jsx
function MyComponent() {
  // data is masked
  const { data } = useSubscription(SUBSCRIPTION, {
    onData: ({ data }) => {
      // data is unmasked
    }
  });
}
```

```jsx
const observable = client.subscribe({ query: SUBSCRIPTION });

observable.subscribe({
  next: ({ data }) => {
    // data is masked
  },
});
```

```jsx
const { subscribeToMore } = useQuery(QUERY);

function startSubscription() {
  subscribeToMore({
    document: SUBSCRIPTION,
    updateQuery: (queryData, { subscriptionData }) => {
      // queryData is unmasked
      // subscriptionData is unmasked
    }
  })
}
```

### Selectively unmasking fragment data

> As you work with data masking more extensively, you may find cases that require access to the full operation result. For such cases, Apollo Client includes an `@unmask` directive that is applied to fragment spreads. Adding `@unmask` to a fragment spread makes the fragment data available.

Reviewer: ```suggestion
As you work with data masking more extensively, you may need access to the full operation result. Apollo Client includes an `@unmask` directive you can apply to fragment spreads. Adding `@unmask` to a fragment spread makes the fragment data available.
```

<Note>

> The `@unmask` directive is an escape hatch. First try adding additional needed fields to your operation before relying on `@unmask`. It is ok however to use `@unmask` to migrate an existing application. Learn more about migrating an existing app in the ["Migrating an existing application" section](#migrating-an-existing-application).

Reviewer: ```suggestion
The `@unmask` directive is an escape hatch. First try adding additional needed fields to your operation before relying on `@unmask`. As an exception, you might use `@unmask` frequently when [migrating an existing application](#migrating-an-existing-application).
```

Author: ```suggestion
The `@unmask` directive is an escape hatch. First try adding additional needed fields to the operation before relying on `@unmask`. As an exception, you might use `@unmask` frequently when [migrating an existing application](#migrating-an-existing-application).
```

I'll raise you a `the` in place of `your` 😛. I couldn't figure out how best to word this, so I appreciate this change!
